platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/hash/MerkleHashBuilder.java

```diff
@@ -142,77 +113,124 @@ public Future<Hash> digestTreeAsync(MerkleNode root) {
-     * @param activeThreadCount
-     * 		If single threaded then this is null, otherwise contains the number of threads
-     * 		that are actively hashing. Once the active thread count dips below the maximum,
-     * 		this means that one thread has either finished or exploded.
      */
-    private void hashSubtree(final Iterator<MerkleNode> it, final AtomicInteger activeThreadCount) {
+    private void hashSubtree(final Iterator<MerkleNode> it) {
         while (it.hasNext()) {
             final MerkleNode node = it.next();
-            // Potential optimization: if this node is currently locked, do not wait here. Skip it and continue.
-            // This would require a lock object that support the "try lock" paradigm.
-            synchronized (node) {
-                if (activeThreadCount != null && activeThreadCount.get() != cpuThreadCount) {
-                    break;
-                }
 
-                if (node.getHash() != null) {
-                    continue;
-                }
+            if (node.getHash() != null) {
+                continue;
+            }
+
+            merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+        }
+    }
+
+    /**
+     * TraverseTask processes the current node in the tree and either hashes it or adds
+     * to a parallel structure of ComputeTasks and creates TraverseTasks for its children.
+     */
+    class TraverseTask extends AbstractTask {
+        final MerkleNode node;
+        final AbstractTask out;
+
+        TraverseTask(final MerkleNode node, AbstractTask out) {
+            super(threadPool, 1);
+            this.node = node;
+            this.out = out;
+        }
 
-                if (node.isLeaf()) {
+        @Override
+        protected boolean exec() {
+            try {
+                if (node == null || node.getHash() != null) {
+                    out.send();
+                } else if (node.isLeaf()) {
                     merkleCryptography.digestSync(node.asLeaf(), MERKLE_DIGEST_TYPE);
+                    out.send();
                 } else {
-                    merkleCryptography.digestSync(node.asInternal(), MERKLE_DIGEST_TYPE);
+                    MerkleInternal internal = node.asInternal();
+                    int nChildren = internal.getNumberOfChildren();
+                    if (nChildren == 0) {
+                        merkleCryptography.digestSync(internal, MERKLE_DIGEST_TYPE);
+                        out.send();
+                    } else {
+                        ComputeTask compute = new ComputeTask(internal, nChildren, out);
+                        for (int childIndex = 0; childIndex < nChildren; childIndex++) {
+                            MerkleNode child = internal.getChild(childIndex);
+                            new TraverseTask(child, compute).send();
+                        }
+                    }
                 }
+            } catch (Exception ex) {
+                out.completeExceptionally(ex);
```

Comment:
If multiple tasks are completed with exceptions, all of them will be propagated to the root (and result) task for no good reason. Besides that, `ResultTask.completeExceptionally()` (which may be called multiple times) calls `future.cancelWithError()`, which may not be called more than once.

Reply:
Done